scipy/_lib/_util.py

```diff
@@ -963,3 +963,97 @@ def _dict_formatter(d, n=0, mplus=1, sorter=None):
                              formatter={'float_kind': _float_formatter_10}):
             s = str(d)
     return s
+
+
+def _apply_over_batch(*argdefs):
+    """
+    Factory for decorator that applies a function over batched arguments.
+
+    Array arguments may have any number of core dimensions (typically 0,
+    1, or 2) and any broadcastable batch shapes. There may be any
+    number of array outputs of any number of dimensions. Assumptions
+    right now - which are satisfied by all functions of interest in `linalg` -
+    are that all array inputs are consecutive keyword or positional arguments,
+    and that the wrapped function returns either a single array or a tuple of
+    arrays. It's only as general as it needs to be right now - it can be extended.
+
+    Parameters
+    ----------
+    *argdefs : tuple of (str, int)
+        Definitions of array arguments: the keyword name of the argument, and
+        the number of core dimensions.
+
+    Example:
+    --------
+    `linalg.eig` accepts two matrices as the first two arguments `a` and `b` and
+    returns one array or a tuple of arrays  depending on the values of other
+    positional or keyword arguments. To generate a wrapper that applies the function
+    over batches of matrices:
+
+    >>> _apply_over_batch(('a', 2), ('b', 2))
```

Comment:
Yup, that's an improvement. Still am not sure how to spell out the difference between one required one optional arg and two required args, but since it's not user-facing and devs are expected to be ready to poke around in code, not just docs, it's fine with https://github.com/scipy/scipy/pull/21462#discussion_r1878399436